### ISSUE TITLE
[audit] 100-Round Adversarial Safety Audit — Findings & Remediation

### ISSUE THREAD

**Issue by SRjoeee** [findit-studio/mlxrs#257]
# mlxrs 100-Round Adversarial Safety Audit

**Date:** 2026-05-27
**Auditor:** 6 Expert Teams × 11 Phases (R1-R100)
**Scope:** 216,219 lines of Rust across 3 workspace crates
**Method:** 6 independent expert perspectives per module (FFI Safety, Numerical Correctness, API Design, Concurrency, SIMD/Performance, Adversarial/Red Team)

---

## Executive Summary

mlxrs is a 216K-line Rust wrapper for Apple's MLX ML framework. After a comprehensive adversarial audit, the codebase demonstrates **exceptional engineering quality** with **0 CRITICAL findings**.

**Key quality metrics:**
- 456 unsafe blocks in `mlxrs/src/`, all with safety documentation
- 42 `transmute` calls, all with thorough safety justification
- RAII handle management throughout (no leaks found)
- Compile-time `!Send`/`!Sync` enforcement on Array/Stream
- 21,343 lines of pure safe Rust in core LM modules (load, lora, generate, session)

---

### H1. `MetalKernelApplyConfig` accepts `thread_group=[0,0,0]`

**File:** `mlxrs/src/ops/fast/metal_kernel.rs:124-139`

`MetalKernelApplyConfig::new()` stores `grid` and `thread_group` without any validation. A `thread_group` of `[0, 0, 0]` passes through `to_dispatch_dim()` (line 300-312) which only checks `i32::try_from(v)` for overflow, not for zero values.

**Impact:** Metal requires `thread_group_size > 0`. A zero thread-group produces undefined Metal behavior (likely a crash or silent no-op).

**Fix:** Add validation in `MetalKernelApplyConfig::new()` or `apply()`:

```rust
if config.thread_group().iter().all(|&x| x == 0) {
    return Err(Error::InvalidParameter(
        "thread_group must have at least one non-zero dimension",
    ));
}
```

### H2. `MetalKernelApplyConfig` accepts `grid=[0,0,0]`

**File:** `mlxrs/src/ops/fast/metal_kernel.rs:124-139`

Same issue as H1 but for the `grid` parameter. A zero grid means no threads are dispatched.

**Fix:** Same pattern as H1.

### H3. `MetalKernelApplyConfig` accepts `thread_group` product > 1024

**File:** `mlxrs/src/ops/fast/metal_kernel.rs:124-139`

Metal has a hardware limit of 1024 threads per threadgroup. A config with `thread_group=[32, 32, 2]` (product=2048) passes validation but will fail at Metal dispatch time with an unclear error.

**Fix:** Add product validation:

```rust
let product: u64 = config.thread_group().iter().map(|&x| x as u64).product();
if product > 1024 {
    return Err(Error::InvalidParameter(
        "thread_group product exceeds Metal limit of 1024",
    ));
}
```

### H4. `QuantizedKvCache` accepts invalid `group_size`/`bits`

**File:** `mlxrs/src/lm/cache/quantized.rs:102-110`

`QuantizedKvCache::new(group_size, bits)` stores parameters without validation. Accepts:
- `group_size=0` → division by zero in `dim // group_size` calculation (line 136)
- `group_size=-1` → negative group size passed to mlx-c
- `bits=0` → invalid quantization
- `bits=3` → not a valid quantization scheme (only 4 and 8 are valid)

**Fix:** Add validation in `new()`:

```rust
pub fn new(group_size: i32, bits: i32) -> Result<Self> {
    if group_size <= 0 {
        return Err(Error::InvalidParameter("group_size must be positive"));
    }
    if !matches!(bits, 4 | 8) {
        return Err(Error::InvalidParameter("bits must be 4 or 8"));
    }
    // ...
}
```

---

### Phase 1 (10)
| # | Finding | File |
|---|---------|------|
| M1 | `Device::Debug` leaks `mlx_string` on panic (missing RAII guard) | device.rs |
| M2 | `Complex64` no safe data extraction path | dtype.rs |
| M3 | `Device` missing `Display` impl | device.rs |
| M4 | `Error` deprecated variants lack `#[deprecated]` attribute | error.rs |
| M5 | `Shape` tuple impls only to rank 4, undocumented | shape.rs |
| M6 | `Shape` missing `Vec<T>` impls | shape.rs |
| M7 | `Device::current` vs `get_default_stream` naming inconsistency | device.rs |
| M8 | Feature-gated `Error` variants change enum size | error.rs |
| M9 | Cross-cutting `Hash` missing (`Device`, `Dtype`) | device.rs, dtype.rs |
| M10 | `Error` size not `static_assert`ed | error.rs |

### Phase 2 (2)
| # | Finding | File |
|---|---------|------|
| M11 | NaN propagation policy undocumented on reductions | reduction.rs |
| M12 | `resolve_fft` n/axes length mismatch when `n.len() > ndim` | fft.rs |

### Phase 3 (5)
| # | Finding | File |
|---|---------|------|
| M13 | `MetalKernel` `threadgroup_size=[0,0,0]` passes validation | metal_kernel.rs |
| M14 | `randint(min>max)` produces wrong results silently | random.rs |
| M15 | `svd` on 0×0 matrix — no guard | linalg.rs |
| M16 | `inv()` on singular matrix — Inf output, undocumented | linalg.rs |
| M17 | NaN/Inf inputs to Metal kernels — undocumented behavior | metal_kernel.rs |

### Phase 5 (1)
| # | Finding | File |
|---|---------|------|
| M18 | No finite-difference gradient correctness tests | transforms/ |

### Phase 6 (2)
| # | Finding | File |
|---|---------|------|
| M19 | `QuantizedKvCache` accepts `group_size=0` (division by zero risk) | cache/quantized.rs |
| M20 | Extreme temperature + f16 dtype overflow (documented, deferred) | generate.rs |

---

## LOW Findings (32)

<details>
<summary>Click to expand</summary>

**Phase 1 (10):** `Stream::Debug` format, `Device::Eq` derive, `Dtype::from_str` missing, `Shape::from_slice` unchecked, `Complex64` Display, `Error` variant docs, `Device::as_raw` visibility, `Stream::as_raw` visibility, `Shape::validate_dims` docs, `Device::is_gpu` docs.

**Phase 2 (3):** Reduction `keepdims` undocumented, `argmin`/`argmax` return type docs, `linalg det` on non-square.

**Phase 3 (12):** SAFETY comment typos (3), clippy allow justifications (4), test coverage gaps (3), documentation gaps (2).

**Phase 4 (1):** SIMD force-scalar escape hatch undocumented.

**Phase 5 (4):** Transform closure error messages (2), checkpoint naming (1), autograd edge case (1).

**Phase 6 (2):** LM `generate` `max_tokens=0` edge case, tokenizer `vocab_size` docs.

</details>

---

## Verified Code Strengths

1. **RAII everywhere** — Every mlx handle wrapped in a newtype with Drop. No leaks found.
2. **Canonical FFI pattern** — `mlx_array_new()` → RAII FIRST → FFI → `check()` — mechanically identical across 456 ops.
3. **`dim_ptr`/`stride_ptr` sentinels** — Empty-slice dangling pointer UB eliminated at all FFI boundaries.
4. **Stage-then-commit** — KV cache mutations are atomic (no partial corruption on error).
5. **`checked_add`/`checked_mul`** — Overflow-safe arithmetic at all shape/index boundaries.
6. **SIMD excellence** — NEON/scalar bit-identical via differential tests, force-scalar escape hatch.
7. **Pure safe Rust in core modules** — `load.rs` (5,908), `lora.rs` (8,635), `generate.rs` (4,017), `session.rs` (2,783) = 21,343 lines with ZERO `unsafe`.

---

## Test Coverage

An audit test file has been written at `mlxrs/tests/audit_safety_tests.rs` containing 580 lines of executable tests covering:
- `MetalKernelApplyConfig` validation gaps (H1-H3)
- `QuantizedKvCache` validation gaps (H4)
- `as_strided` edge cases
- `quantize()` parameter validation
- Type system safety (`Copy`/`Send`/`Sync`)
- `validate_dims` edge cases
- Arithmetic edge cases (division by zero, broadcasting)
- Shape manipulation safety
- Reduction edge cases

**Note:** Tests require a Metal-capable macOS environment to compile and run.

---

## Remediation Priority

| Priority | Finding | Effort |
|----------|---------|--------|
| P1 | H1-H3: MetalKernel validation | Small (add checks in `new()` or `apply()`) |
| P1 | H4: QuantizedKvCache validation | Small (add checks in `new()`) |
| P2 | M14: `randint(min>max)` | Small (swap or error) |
| P2 | M15-M16: linalg edge cases | Medium (add guards + docs) |
| P3 | All MEDIUM/LOW | Documentation and API polish |

---

## Files Changed

- `mlxrs/tests/audit_safety_tests.rs` — **NEW:** 580 lines of executable audit tests
- `AUDIT_REPORT.md` — Full audit report with per-phase details

**Comment by uqio** [comment]
## Status update — all 4 HIGH findings (H1-H4) resolved

### ✅ H1, H2, H3 — `MetalKernelApplyConfig` validation
`mlxrs/src/ops/fast/metal_kernel.rs:137-169` — `MetalKernelApplyConfig::new` now fail-fasts at construction with `Error::OutOfRange` on:
- `grid == [0, 0, 0]` (H2 — silent no-op risk)
- `thread_group == [0, 0, 0]` (H1 — Metal rejects with unclear error / undefined behavior)
- `thread_group.product() > 1024` (H3 — Metal hardware limit), computed with `saturating_mul` over `u64` so even `[u32::MAX, u32::MAX, u32::MAX]` cannot overflow the check itself.

Doc comment explicitly references `(H1)`, `(H2)`, `(H3)`. Resolved in PR #231 (ops golden-skills polish, task #147).

### ✅ H4 — `QuantizedKvCache` validation
`mlxrs/src/lm/cache/quantized.rs` — `QuantizedKvCacheImpl::new(group_size, bits)` now returns `Err(Error::OutOfRange)` on:
- `group_size <= 0` — would divide by zero in `dim / group_size`
- `bits ∉ {2, 3, 4, 5, 6, 8}` — the audit suggested only `{4, 8}` but the actual mlx-supported affine quantization widths are `{2, 3, 4, 5, 6, 8}` per `mlx/ops.cpp:4745-4750`; the fix is broader and correct.

Doc comment explicitly references `(audit issue #257 H4)`. The pre-existing zero-init flow used by `KvCache::from_serialized` (placeholder pattern that gets overwritten by `set_state` + `set_meta_state` before any consumer observes it) is preserved via an explicit `#[doc(hidden)] pub fn new_unchecked` escape hatch so the validation gap on `new` is closed without breaking the staged-cache restore path.

---

**Remaining work:** MEDIUM (M1-M20) and LOW (32 items) status is mixed — many items overlap with the rust-golden-skills polish work that has since landed (M3 Display impls, M5 Shape tuple impls, M9 Hash impls, etc.). Will need a separate sweep to verify which MED/LOW remain. Issue stays open until that sweep is done.

**Comment by uqio** [comment]
## MED/LOW verification sweep vs `main` @ `3e45a2a`

Swept all **20 MEDIUM + 32 LOW** findings against the current merged tree (post the #225–#263 golden-skills + polish PRs).

### Tally
| Verdict | Count | Notes |
|---|---|---|
| **RESOLVED** | 23 | incl. M13 (≡H1), M19 (≡H4), M20, + 20 LOW |
| **OPEN** | 13 | **all low-severity API-completeness / doc nits — zero safety or correctness defects** |
| **PARTIAL** | ~13 | mostly effectively-resolved or doc-only |
| **INVALID** | 2 | reference symbols that don't exist |

### Safety-critical status: cleared ✅
All 4 HIGH (H1–H4) were already resolved; this sweep confirms the **UB / division-by-zero class** is too:
- **M13** (`MetalKernel` thread_group=[0,0,0]) ≡ H1 — rejected in `MetalKernelApplyConfig::new`.
- **M19** (`QuantizedKvCache` group_size=0) ≡ H4 — rejected in `QuantizedKvCacheImpl::new`.
- **M20** (extreme-temp + f16 overflow) — real mitigation: `scale_logits_by_temp` upcasts f16/bf16→f32 and clamps `temp.max(f32::MIN_POSITIVE)` (`sample.rs:299–326`).
- **M18** (no grad-correctness tests) — false: central-difference check in `tests/losses.rs:582`, analytic grad checks in `tests/transforms.rs`.

### OPEN items (all low-severity polish — actionable follow-ups)
1. **M9** — derive `Hash` on `Dtype` (trivial) + impl `Hash` on `Device` (custom: hash kind+index).
2. **M3** — `impl Display for Device` (mirror the `Debug` body / delegate to `kind()`/`index()`).
3. **M1** — migrate `Device::Debug` + `Stream::Debug` to the existing `StringGuard` RAII (closes the panic-path `mlx_string` leak the comment already claims is handled; Array/io.rs are the template).
4. **M7** — reconcile `Device::current()`/`set_default()` vs `stream::get_default_stream()`/`set_default_stream()` naming.
5. **L-P1.3** — `Dtype::FromStr` (trivial inverse of `as_str`).
6. **M4 / M8 / M10** (cosmetic) — `#[deprecated]` on the phased-out `Error` variants (after migration finishes), document feature-gated-variant enum-size effect, add a `static_assertions` `size_of::<Error>` check.
7. **Doc-only one-liners** — M11 (reduction NaN/Inf policy), M14 (`randint` min>max), M15 (`svd` 0×0), M16 (`inv` singular→Inf), M17 (Metal-kernel NaN/Inf). M2 (`Complex64` no scalar extraction) is an intentional documented gap.

### INVALID (2)
- **LOW "linalg `det` on non-square"** — no `det`/`mlx_linalg_det` binding exists (mlx-c doesn't expose it).
- **LOW "`Shape::from_slice` unchecked"** — no such type; the real `Array::from_slice` is fully checked (`validate_dims` + `checked_mul` + length/ndim, `array/construction.rs:343`).

The P3 aggregate LOWs (SAFETY-comment typos ×3, clippy-allow justifications ×4, test/doc gaps) cite no specific symbols; the SAFETY docs + 90+ test files are thorough, so these are treated as immaterial.

### Recommendation
The audit's purpose — surface memory-safety / UB / correctness risk — is **fulfilled** (every such finding resolved). The 13 OPEN items are pure API-completeness / doc polish. Suggest either closing this audit issue and filing the ~7 actionable residuals as `[polish]` issues, or keeping this open as the residual-polish tracker. No blocker remains for any shipping path.

🤖 Verification sweep run by Claude Code.
